Get scalar data from clients with old pings

Dexterp37 opened the pull request:
The scalars section was added by bug 1282091, which
landed on July 2016. Pings before that date don't
have the scalars section. If we have a ping with
the missing section among the pings for a client,
the longitudinal view drops all the data from that
client, if if newer pings have the scalar section.
This patch prevents that from happening.
Bonus: this also adds some documentation about
running this code on Windows.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/telemetry-batch-view/124)
<!-- Reviewable:end -->

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

Dexterp37 commented:
@vitillo , with this change if a particular ping doesn't have a specific scalar or a user doesn't send a scalar, it gets set to a default value (0 for uint, false for boolean and "" for strings). This is not desirable. Any idea about how to allow null values in the schema and write them if no scalar is available?

vitillo commented:
Yeah, that's pretty bad as it could cause analyst to reach wrong conclusions. You should be able to store `null` values within the arrays, have you tried that?

vitillo commented on README.md:
````diff
@@ -43,3 +43,25 @@ If you run into memory issues during compilation time issue the following comman
 export _JAVA_OPTIONS="-Xss4M -Xmx2G"
 ```
+
+**Running on Windows**
````
This change is not logically related to the rest of the patch. Please split the PR into multiple commits.

Dexterp37 commented:
Yeah, I tried by doing that:

- Changed the schema definition for the UintScalar to `builder.name(key).`type`().optional().array().items().nullable().longType()` (~line 511).
- Changed the default values for missing scalars to `null` or `None`, i.e. `vectorizeScalar_(name, payloads, s => s.asInstanceOf[BigInt], null)` (~line 737).

This doesn't seem to work, as it tries to cast the `null`/`None` to a Number and raises an exception:

> java.lang.ClassCastException: scala.None$ cannot be cast to java.lang.Number

vitillo commented:
Where is that exception thrown?

Dexterp37 commented:
@vitillo, here's the full exception, triggered by `private val path = ParquetFile.serialize(List(record).toIterator, schema):` from LongitudinalTest.scala.

> An exception or error caused a run to abort: scala.None$ cannot be cast to java.lang.Number 
java.lang.ClassCastException: scala.None$ cannot be cast to java.lang.Number
	at org.apache.parquet.avro.AvroWriteSupport.writeValue(AvroWriteSupport.java:202)
	at org.apache.parquet.avro.AvroWriteSupport.writeArray(AvroWriteSupport.java:127)
	at org.apache.parquet.avro.AvroWriteSupport.writeValue(AvroWriteSupport.java:216)
	at org.apache.parquet.avro.AvroWriteSupport.writeRecordFields(AvroWriteSupport.java:112)
	at org.apache.parquet.avro.AvroWriteSupport.write(AvroWriteSupport.java:87)
	at org.apache.parquet.avro.AvroWriteSupport.write(AvroWriteSupport.java:44)
	at org.apache.parquet.hadoop.InternalParquetRecordWriter.write(InternalParquetRecordWriter.java:121)
	at org.apache.parquet.hadoop.ParquetWriter.write(ParquetWriter.java:258)
	at com.mozilla.telemetry.parquet.ParquetFile$$anonfun$serialize$2.apply(ParquetFile.scala:31)
	at com.mozilla.telemetry.parquet.ParquetFile$$anonfun$serialize$2.apply(ParquetFile.scala:30)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at com.mozilla.telemetry.parquet.ParquetFile$.serialize(ParquetFile.scala:30)
	at com.mozilla.telemetry.views.LongitudinalTest$$anon$1.<init>(LongitudinalTest.scala:205)
	at com.mozilla.telemetry.views.LongitudinalTest.<init>(LongitudinalTest.scala:182)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.scalatest.tools.Runner$.genSuiteConfig(Runner.scala:2644)
	at org.scalatest.tools.Runner$$anonfun$37.apply(Runner.scala:2461)
	at org.scalatest.tools.Runner$$anonfun$37.apply(Runner.scala:2460)
	at scala.collection.immutable.List.map(List.scala:273)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2460)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
	at org.scalatest.tools.Runner$.run(Runner.scala:883)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:138)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)

vitillo commented:
@Dexterp37 Alright, this can't be checked in as is so you would have to understand if this is a limitation of Avro, Parquet or a bug in their respective implementation. If that's the case, we might have to fall back to another approach, like converting the integers to floats and using nans.

Dexterp37 commented:
You, thanks Roberto. I'll sort this issue out!

Dexterp37 commented:
I think I've figured the problem. Apparently, in order to support Null values in arrays, one should in theory simply make the changes that I suggested in [this comment](https://github.com/mozilla/telemetry-batch-view/pull/124#issuecomment-253545242), by replacing for example

`builder.name(key).`type`().optional().array().items().longType()`

to 

`builder.name(key).`type`().optional().array().items().nullable().longType()`

to allow null values, and then writing `nulls` if no scalar is found. `SchemaBuilder` seems to produce the right schema, as the Avro definition becomes (taken from a scalar from the tests):

>   {
>     "name" : "scalar_parent_mock_scalar_uint",
>     "type" : [ "null", {
>       "type" : "array",
>       "items" : [ "long", "null" ]
>     } ],
>     "default" : null
>   }
  
The "items" entry is a Union type between Longs and Nulls, as expected. When `AvroSchemaConverter` converts this to a Parquet-friendly schema, it generates

>   optional group scalar_parent_mock_scalar_uint (LIST) {
>     repeated int64 array;
>   }
  
using the `repeated` keyword instead of `optional`, thus requiring non-null longs in the scalar and failing to write Parquet files.

This issue is addressed by [PARQUET-285](https://issues.apache.org/jira/browse/PARQUET-285) available in parquet-avro 1.8.0, by using a 3-level structure for list, documented [here ](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md). By using that version of *org.apache.avro* and *org.apache.parquet*, the file is written correctly with `null` values in the array of scalar values. Here's the correct schema generated by the 1.8.0 version:

>   optional group scalar_parent_mock_scalar_uint (LIST) {
>     repeated group list {
>       optional int64 element;
>     }
>   }

Unfortunately, this format change for this very specific use case breaks Spark parquet file loading. A fix is discussed in [SPARK-16344](https://issues.apache.org/jira/browse/SPARK-16344) and will make it to Spark 2.1.0.

Dexterp37 commented:
By nesting the scalar data inside a bogus structure, we manage to get the expected schema layout, thus we're able to use nulls.

>   optional group scalar_parent_mock_scalar_uint (LIST) {
>     repeated group array {
>       optional int64 value;
>     }
>   }

We should probably still change that once we move to Spark 2.1.0. I'm currently testing that on a Spark cluster.

Dexterp37 commented:
This seems to behave correctly, as it shows the [null]s instead of chopping off the data.

> scala> val tst = spark.sql("select scalar_parent_browser_engagement_max_concurrent_tab_count as the_scalar FROM scalars_new").where("scalar_parent_browser_engagement_max_concurrent_tab_count is not null")
> tst: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [the_scalar: array<struct<value:bigint>>]
> 
> scala> tst.show
> +--------------------+
> |          the_scalar|
> +--------------------+
> |[[2], [2], [1], [1]]|
> |          [[1], [1]]|
> |[[1], [7], [8], [...|
> |[[5], [1], [1], [...|
> |               [[8]]|
> |[[1], [2], **[null].**..|
> |[[14], [4], [5], ...|
> |     [[3], [4], [3]]|
> |               [[6]]|
> |[[1], [4], [4], [...|
> |[[11], [7], [2], ...|
> |     [[1], [4], [1]]|
> |     [[5], [1], [1]]|
> |          [[1], [2]]|
> |[[9], [1], [7], [...|
> |[[8], [6], [1], [...|
> |[[1], [1], [1], [...|
> |[[14], [3], [1], ...|
> |[[1], [4], [2], [...|
> |       [[2], [null]]|
> +--------------------+
> only showing top 20 rows

@vitillo , this seems ready for a review :-D

vitillo commented:
@Dexterp37 In your example query, it seems like each individual scalar value ends up in an array; shouldn't it end up in a record?

vitillo commented:
Looks good, thanks!
